File: src/App/Platforms/Android/Tiles/GeneratorTileService.cs
using Android.App;
using Android.Content;
using Android.Runtime;
using Android.Service.QuickSettings;
using Bit.App.Droid.Utilities;
using Java.Lang;

namespace Bit.Droid.Tile
{
    [Service(Permission = Android.Manifest.Permission.BindQuickSettingsTile, Exported = true, Label = "@string/PasswordGenerator",
        Icon = "@drawable/generate")]
    [IntentFilter(new string[] { ActionQsTile })]
    [Register("com.x8bit.bitwarden.GeneratorTileService")]
    public class GeneratorTileService : TileService
    {
        public override void OnTileAdded()
        {
            base.OnTileAdded();
        }

        public override void OnStartListening()
        {
            base.OnStartListening();
        }

        public override void OnStopListening()
        {
            base.OnStopListening();
        }

        public override void OnTileRemoved()
        {
            base.OnTileRemoved();
        }

        public override void OnClick()
        {
            base.OnClick();

            if (IsLocked)
            {
                UnlockAndRun(new Runnable(() =>
                {
                    LaunchMyVault();
                }));
            }
            else
            {
                LaunchMyVault();
            }
        }

        private void LaunchMyVault()
        {
            var intent = new Intent(this, typeof(MainActivity));
            intent.SetFlags(ActivityFlags.NewTask | ActivityFlags.SingleTop | ActivityFlags.ClearTop);
            intent.PutExtra("generatorTile", true);
            this.StartActivityAndCollapseWithIntent(intent, isMutable: false);
        }
    }
}

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/tiles/GeneratorTileService.kt
package com.x8bit.bitwarden.ui.platform.feature.tiles

import android.annotation.SuppressLint
import android.os.Build
import android.service.quicksettings.TileService
import com.x8bit.bitwarden.ui.platform.manager.intent.IntentManager
import com.x8bit.bitwarden.ui.platform.manager.intent.IntentManagerImpl
import kotlinx.coroutines.Runnable

class GeneratorTileService: TileService() {
    private val tileIntentManager: IntentManager = IntentManagerImpl(this)

    override fun onTileAdded() {
        super.onTileAdded()
    }

    override fun onStartListening() {
        super.onStartListening()
    }

    override fun onStopListening() {
        super.onStopListening()
    }

    override fun onTileRemoved() {
        super.onTileRemoved()
    }

    override fun onClick() {
        super.onClick()

        if (isLocked) {
            unlockAndRun(Runnable { launchGenerator(); })
        } else {
            launchGenerator()
        }
    }

    @Suppress("DEPRECATION")
    @SuppressLint("StartActivityAndCollapseDeprecated")
    fun launchGenerator() {
        val intent = tileIntentManager.createTileIntent("bitwarden://password_generator");

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            startActivityAndCollapse(tileIntentManager.createTilePendingIntent(intent))
        } else {
            startActivityAndCollapse(intent)
        }
    }
}
